plugins/hls-cabal-plugin/src/Ide/Plugin/Cabal/Outline.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE CPP                   #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE ViewPatterns          #-}
+
+module Ide.Plugin.Cabal.Outline where
+
+import           Control.Monad.IO.Class
+import           Data.Maybe
+import qualified Data.Text                         as T
+import           Data.Text.Encoding                (decodeUtf8)
+import           Development.IDE.Core.Rules
+import           Development.IDE.Core.Shake        (IdeState (shakeExtras),
+                                                    runIdeAction,
+                                                    useWithStaleFast)
+import           Development.IDE.Types.Location    (toNormalizedFilePath')
+import           Distribution.Fields.Field         (Field (Field, Section),
+                                                    Name (Name),
+                                                    SectionArg (SecArgName, SecArgOther, SecArgStr))
+import           Distribution.Parsec.Position      (Position)
+import           Ide.Plugin.Cabal.Completion.Types (ParseCabalFields (..),
+                                                    cabalPositionToLSPPosition)
+import           Ide.Plugin.Cabal.Orphans          ()
+import           Ide.Types                         (PluginMethodHandler)
+import           Language.LSP.Protocol.Message     (Method (..))
+import           Language.LSP.Protocol.Types       (DocumentSymbol (..))
+import qualified Language.LSP.Protocol.Types       as LSP
+
+
+moduleOutline :: PluginMethodHandler IdeState Method_TextDocumentDocumentSymbol
+moduleOutline ideState _ LSP.DocumentSymbolParams {_textDocument = LSP.TextDocumentIdentifier uri} =
+  case LSP.uriToFilePath uri of
+    Just (toNormalizedFilePath' -> fp) -> do
+      mFields <- liftIO $ runIdeAction "cabal-plugin.fields" (shakeExtras ideState) (useWithStaleFast ParseCabalFields fp)
+      case fmap fst mFields of
+        Just fieldPositions -> pure $ LSP.InR (LSP.InL allSymbols)
+          where
+            allSymbols = mapMaybe documentSymbolForField fieldPositions
+        Nothing -> pure $ LSP.InL []
+    Nothing -> pure $ LSP.InL []
+
+-- | Creates a DocumentSumbol object for the
+--   cabal AST, without displaying fieldLines and
+--   displaying Section name and SectionArgs in one line
```

Comment:
Could you expand on this with a concrete example? E.g. 


We turn an cabal ast such as
```
-- @
--   library
--     build-depends: base
-- @
-- 
-- into 
-- 
-- @ 
--   library
--     build-depends
-- @ 
-- 
-- omitting the value of the field `build-depends`, as it scaled poorly on big projects.